### ISSUE TITLE
Writing Logs via jsonPayload is undocumented

### ISSUE THREAD

**Issue by geigerj** [GoogleCloudPlatform/google-cloud-ruby#1521]
## Context

This is ported from an internal report from @palladius. @quartzmo @blowmage Could you advise on how to use `jsonPayload` with the Ruby client library?

### Issue summary 

As a Cloud Logging developer, I'm able to use (as described in [1]) either of  "textPayload",  "jsonPayload" or  "protoPayload": I was able to test first 2 in API Explorer. However, i'm unable to use jsonPayload from client libraries. 

### Request

API Request: https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry

```
def logga_on_cloud_json(logging, mystr)
        entry = logging.entry
        entry.log_name = "whatevs"
        entry.severity = :NOTICE
        entry.resource.type = "global"
        #p entry.methods()
        #entry.json_payload = "[#{$short_hostname}] #{mystr}" #
        #entry.resource.labels[:version_id] = VER
        #entry.resource.labels[:module_id] = Socket.gethostname
        my_json_string = '{"desc":{"someKey":"someValue","anotherKey":"value"},"main_item":{"stats":{"a":8,"b":12,"c":10}}}'
        entry.payload = my_json_string  
        # Writes the log entry
        logging.write_entries entry
end
```

### API Response

My json string is treated as a normal string, and I've tried using "entry.json_payload =" and it fails. Only entry.payload  is accepted, how do I make the client library aware I want to give JSON or proto?

In other words, I'd like to be able to programmatically forge this payload:

```
POST https://logging.googleapis.com/v2/entries:write?key={YOUR_API_KEY}

{
 "entries": [
  {
   "jsonPayload": {
    "foo": "baz",
    "messaggio": "my message"
   },
   "resource": {
    "type": "global"
   },
   "logName": "projects/gae-debiti/logs/gic-logga"
  }
 ]
}
```

## Next steps

1. Please provide steps on how to use *programmatically* JSON/proto
2. Amend docs with sample code please. 

[1] https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry

**Comment by quartzmo** [comment]
I believe this issue is correct: I could not find any documentation examples showing the use of a JSON payload. I did however find this [acceptance test](https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-logging/acceptance/logging/logging_test.rb#L130).

**Comment by quartzmo** [comment]
As shown in the acceptance test, the [supported way to set a JSON value](https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-logging/lib/google/cloud/logging/entry.rb#L463-L464) as the payload is to pass a hash to `Entry#payload=` (not `#json_payload=`).

**Comment by blowmage** [comment]
I don't believe this is undocumented. The docs for [`Entry.payload=`](https://googlecloudplatform.github.io/google-cloud-ruby/#/docs/google-cloud-logging/v1.1.0/google/cloud/logging/entry?method=payload%3D-instance) state:

> The log entry payload, represented as either a string, a hash (JSON), or a hash (protocol buffer).

So to change your code use `jsonPayload` don't convert the JSON value to a string, convert it to a Hash:

```ruby
def logga_on_cloud_json(logging, mystr)
        entry = logging.entry
        entry.log_name = "whatevs"
        entry.severity = :NOTICE
        entry.resource.type = "global"
        #p entry.methods()
        #entry.json_payload = "[#{$short_hostname}] #{mystr}" #
        #entry.resource.labels[:version_id] = VER
        #entry.resource.labels[:module_id] = Socket.gethostname
        entry.payload = {"desc"=>{"someKey"=>"someValue", "anotherKey"=>"value"},
                         "main_item"=>{"stats"=>{"a"=>8, "b"=>12, "c"=>10}}}
        # Writes the log entry
        logging.write_entries entry
end
```

**Comment by quartzmo** [comment]
@blowmage Not undocumented, but there is no example.

**Comment by palladius** [comment]
So it looks like the culprit is this: https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-logging/lib/google/cloud/logging/entry.rb#L461-L466

You add content to entry.payload. Then:
1. If it belongs to class Google::Protobuf::Any ==> protobuf
2. Else if it is hashable (responds to :to_hash) ==> JSON
3. Else convert to string.

I created a has payload and it works like a charm.

		logging = Google::Cloud::Logging.new project: project_id
		entry = logging.entry
		hash_payload = {
			 "header": {
			 	"hostname": short_hostname, 
			 	"domain": domain_name,
				"project_id": project_id,
				"script_ver": VER,
				"ruby_version": RUBY_VERSION,
				"username": ENV['USER'],
			   },
			 "payload": mystr,
		}
		#entry.payload = "[#{short_hostname}] #{mystr}"  # used to be a string
		entry.payload = hash_payload # responds to hash
                logging.write_entries entry

**Comment by quartzmo** [comment]
What changes are needed to resolve this issue, if any?

**Comment by frankyn** [comment]
Following-up from internal bug:

The change was to document Hash payloads when logging entries. Currently `Entry#payload=` doesn't show an example of how to do this. 

![fhhcechb4rz](https://user-images.githubusercontent.com/1907648/28685820-ed4e67e8-72bd-11e7-9b82-1914666ed0d4.png)

There is an example for String payloads, but no Hash payload examples from 

![r3lklk47csm](https://user-images.githubusercontent.com/1907648/28685901-315c0aa8-72be-11e7-9dc3-63a032e3c314.png)

A stretch goal is to add an example of using a Protobuf payload, but I'm not sure what a concise example could look like.

**Comment by Nakilon** [comment]
@palladius where did you get this structure from? It reminds me objects that I saw in Error Reporting but can't find docs by googling `"script_ver"`
